python_modules/libraries/dagster-dask/dagster_dask/resources.py

```diff
@@ -0,0 +1,143 @@
+from dagster import (
+    Bool,
+    Field,
+    Int,
+    Permissive,
+    Selector,
+    Shape,
+    String,
+    resource,
+)
+from dask.distributed import Client
+
+
+DaskClusterTypes = {
+    key: dict(zip(("name", "module", "class"), values))
+    for key, values in {
+        "local": ("Local", "dask.distributed", "LocalCluster"),
+        "yarn": ("YARN", "dask_yarn", "YarnCluster"),
+        "ssh": ("SSH", "dask.distributed", "SSHCluster"),
+        "pbs": ("PBS", "dask_jobqueue", "PBSCluster"),
+        "moab": ("Moab", "dask_jobqueue", "MoabCluster"),
+        "sge": ("SGE", "dask_jobqueue", "SGECluster"),
+        "lsf": ("LSF", "dask_jobqueue", "LSFCluster"),
+        "slurm": ("SLURM", "dask_jobqueue", "SLURMCluster"),
+        "oar": ("OAR", "dask_jobqueue", "OARCluster"),
+        "kube": ("Kubernetes", "dask_kubernetes", "KubeCluster"),
+    }.items()
+}
+
+
+class DaskResource(object):
+    def __init__(self, context):
+        # Create a Dask cluster if a cluster config is specified.
+        # This will be passed as the address value to the Client.
+        cluster_config = context.resource_config.get("cluster", None)
+        if cluster_config:
+            from importlib import import_module
+
+            cluster_type, cluster_opts = next(iter(cluster_config.items()))
+            cluster_meta = DaskClusterTypes.get(cluster_type, None)
+            if not cluster_meta:
+                raise ValueError(f"Unknown cluster type “{cluster_type}”.")
+
+            # Import the cluster module by name, and get the cluster Class.
+            cluster_module = import_module(cluster_meta["module"])
```

Comment:
a bit concerned about the late import here, and how missing dependency failures will manifest - but I can't come up with a solution that really solves or that. This same problem exists in the executor as well.

Reply:
Few ideas:

1. Each cluster type's field description could mention their required module.
2. Is there a hook for validating resource configs? That might be a good time to check if a module exists.
3. The `dagster_dask` module could declare `extra_requires` for the modules. Though, I'm not sure if this is something this module wants to take on, considering Dask's `distributed` module doesn't already do it, and it might not be obvious to users.

Reply:
As an aside, I'm aware that the resource and executor will have separate implementations for configuring and creating Dask clients and clusters with this PR. I'm thinking of following up with another PR later to bring them together under a common implementation.

Reply:
I've added the module name to each cluster config's description. For example, the `kube` type's description now says:

> Kubernetes cluster config. Requires dask_kubernetes.